### PR TITLE
Support url parsing at compile-time via proc-macro

### PR DESCRIPTION
**Pull request by NobodyXu**
Added new crate `url_macro`, usage:

```rust
const URL: url::Url = url_macro::parse!("https://www.github.com");
assert_eq!(URL, url::Url::parse("https://www.github.com").unwrap());
```

# Motivation

It is not uncommon for users to want to parse `Url` at compile time, instead of having to use `OnceCell`/`OnceLock`, it would be more ergonomic/efficient to parse it at compile-time.

### CODE REVIEW

File: url/src/lib.rs [hunk 1]
```diff
@@ -194,7 +194,7 @@ pub struct Url {
     ///   authority = "//" userinfo? host [ ":" port ]?
     ///   userinfo = username [ ":" password ]? "@"
     ///   hierarchical-path = [ "/" path-segment ]+
-    serialization: String,
+    serialization: Cow<'static, str>,
```
Comment (NobodyXu) [inline url/src/lib.rs:197]: Could replace this with [`beef::lean::Cow`](https://docs.rs/beef/latest/beef/lean/type.Cow.html) to reduce size of `Url`, but might be a pre-mature optimization

File: url-macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/// Parse url at compile-time
+///
+/// ```
+/// const URL: url::Url = url_macro::parse!("https://www.github.com");
+/// # fn main() {
+/// assert_eq!(URL, url::Url::parse("https://www.github.com").unwrap());
+/// # }
+/// ```
+#[proc_macro]
+pub fn parse(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
```
Comment (NobodyXu) [inline url-macro/src/lib.rs:25]: TODO: Add test for `parse!`

**Review by Manishearth — CHANGES_REQUESTED** [review]
I'm not convinced this is worth the addition of the Cow, or worth it for _this_ project to maintain

**Comment by NobodyXu** [comment]
Thanks for the quick feedback, in that case I will close this.

**Pull request closed without merging**